version: '3.8'

volumes:
  pg_data_test:

networks:
  dev-test:
    driver: bridge

services:
  db_test:
    image: postgres:15.1-alpine
    container_name: postgres_test
    restart: always
    networks:
      dev-test:
    volumes:
      - pg_data_test:/var/lib/postgresql/data
    environment:
      POSTGRES_DB: ${TEST_DB_NAME}
      POSTGRES_USER: ${TEST_DB_USER}
      POSTGRES_PASSWORD: ${TEST_DB_PASSWORD}
    env_file:
      - .env
    expose:
      - "5432"

    healthcheck:
      test: [ "CMD-SHELL", "sh -c 'pg_isready -U ${TEST_DB_USER} -d ${TEST_DB_NAME}'" ]
      interval: 10s
      timeout: 3s
      retries: 3

  redis_cache_test:
    image: redis:7.2-rc2
    container_name: redis_test
    networks:
      dev-test:
    environment:
      REDIS_PASSWORD: ${TEST_REDIS_PASSWORD}
    volumes:
      - ./redis/redis.conf:/usr/local/etc/redis/redis.conf
    command: /bin/sh -c "redis-server --requirepass $$TEST_REDIS_PASSWORD"
    restart: on-failure
    env_file:
      - .env
    expose:
      - "6379"
    healthcheck:
      test: [ "CMD-SHELL", "redis-cli ping | grep PONG" ]
      interval: 1s
      timeout: 3s
      retries: 5

  app_test:
    build:
      dockerfile: Dockerfile-test
    container_name: app_test
    networks:
      dev-test:
    env_file:
      - .env
    depends_on:
      - db_test
      - redis_cache_test
    ports:
      - "8006:8000"
    volumes:
      - ./:/code
